taichi/ir/frontend_ir.h

```diff
@@ -446,31 +446,39 @@ class ExternalTensorExpression : public Expression {
   int arg_id;
   int element_dim;  // 0: scalar; 1: vector (SOA); 2: matrix (SOA); -1: vector
                     // (AOS); -2: matrix (AOS)
+  bool is_grad;
 
   ExternalTensorExpression(const DataType &dt,
                            int dim,
                            int arg_id,
-                           int element_dim) {
-    init(dt, dim, arg_id, element_dim);
+                           int element_dim,
+                           bool is_grad = false) {
+    init(dt, dim, arg_id, element_dim, is_grad);
   }
 
   ExternalTensorExpression(const DataType &dt,
                            int dim,
                            int arg_id,
                            int element_dim,
-                           const std::vector<int> &element_shape) {
+                           const std::vector<int> &element_shape,
+                           bool is_grad = false) {
     if (element_shape.size() == 0) {
-      init(dt, dim, arg_id, element_dim);
+      init(dt, dim, arg_id, element_dim, is_grad);
     } else {
       TI_ASSERT(dt->is<PrimitiveType>());
 
       auto tensor_type =
           taichi::lang::TypeFactory::get_instance().create_tensor_type(
               element_shape, dt);
-      init(tensor_type, dim, arg_id, element_dim);
+      init(tensor_type, dim, arg_id, element_dim, is_grad);
     }
   }
 
+  ExternalTensorExpression(Expr *expr, bool is_grad = true) {
```

Comment:
Because [here](https://github.com/taichi-dev/taichi/blob/094c05a56a0cfe7f5b3c0d00525b48e3238e699d/python/taichi/lang/any_array.py#L43) we want to construct a new grad ExternalTensorExpression by just using its primal.